lib/Dialect/FIRRTL/Transforms/Dedup.cpp

```diff
@@ -1340,23 +1373,43 @@ class DedupPass : public DedupBase<DedupPass> {
         std::pair<std::array<uint8_t, 32>, SmallVector<StringAttr>>>>
         hashesAndModuleNames(modules.size());
     StructuralHasherSharedConstants hasherConstants(&getContext());
+
+    auto failed = false;
+
     // Calculate module information parallelly.
     mlir::parallelFor(context, 0, modules.size(), [&](unsigned idx) {
       auto module = modules[idx];
+      AnnotationSet annotations(module);
       // If the module is marked with NoDedup, just skip it.
-      if (AnnotationSet(module).hasAnnotation(noDedupClass))
+      if (annotations.hasAnnotation(noDedupClass))
         return;
+
       // If the module has input RefType ports, also skip it.
       if (llvm::any_of(module.getPorts(), [&](PortInfo port) {
             return type_isa<RefType>(port.type) && port.isInput();
           }))
         return;
 
       StructuralHasher hasher(hasherConstants);
+      llvm::SmallSetVector<StringAttr, 1> groups;
+      for (auto annotation : annotations) {
+        if (annotation.getClass() == dedupGroupClass)
+          groups.insert(annotation.getMember<StringAttr>("group"));
+      }
+      if (groups.size() > 1) {
+        module.emitError("module belongs to multiple dedup groups: ") << groups;
+        failed = true;
```

Comment:
I think this is not thread safe. Can we change outer `parallelFor` to `failableParalleFor` and return `failure()` here?